### PR TITLE
fix assisted decoding

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi @gante . This PR is to fix the assisted decoding when the model and assistant model are on different devices.

It can be easily reproduced by:
```python
model = model.to("cuda")
model.generate(**inputs, assistant_model=assistant_model.to("cpu"))
```

### CODE REVIEW

**Comment by jiqing-feng** [comment]
The failed CIs seem not related to my changes

**Review by gante — COMMENTED** [review]
Hi @jiqing-feng! Thank you for opening this PR 🤗 

To the best of my knowledge, the changes you're suggesting should not be needed. As such, I've asked a few questions below to understand why we need these changes :)

File: src/transformers/generation/logits_process.py
Comment (jiqing-feng) [reply src/transformers/generation/logits_process.py:157]: Yes, but before we move the `input_tensor` to the same device as `assiatant_model`, the logits processor already initialized [here](https://github.com/gante/transformers/blob/b7672826cad31e30319487af876e608d8af7d37b/src/transformers/generation/utils.py#L1801). The `eos_token_id` also in the target model device, see [here](https://github.com/huggingface/transformers/blob/d9daeff2978eda167c7256b2c928941f2248bfbe/src/transformers/generation/utils.py#L1477).

**Comment by jiqing-feng** [comment]
Hi @gante . Sorry for not making it clear. Could you run this script:
```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM


model_id = "meta-llama/Llama-2-7b-chat-hf"
assistant_model_id = "Felladrin/Llama-68M-Chat-v1"
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token_id = tokenizer.eos_token_id

model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.float16).to("cuda")
assistant_model = AutoModelForCausalLM.from_pretrained(assistant_model_id, torch_dtype=torch.bfloat16).to("cpu")

prompt = "Assisted decoding is"
inputs = tokenizer(prompt, return_tensors="pt")

model.generate(**inputs, assistant_model=assistant_model, max_new_tokens=8, do_sample=False)
```

It will get the error `Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`.

File: src/transformers/generation/utils.py
Comment (jiqing-feng) [reply src/transformers/generation/utils.py:2186]: The eos_token_id has the same device as the target model, see [here](https://github.com/huggingface/transformers/blob/d9daeff2978eda167c7256b2c928941f2248bfbe/src/transformers/generation/utils.py#L1477).

**Comment by jiqing-feng** [comment]
HI @gante . I just found the real issue happens [here](https://github.com/gante/transformers/blob/b7672826cad31e30319487af876e608d8af7d37b/src/transformers/generation/utils.py#L1470-L1471), pls take a review. Thx!

**Comment by jiqing-feng** [comment]
I would like to add a test for this. Do you know where I should add this test? Thx!

**Review by gante — APPROVED** [review]
This makes sense, thank you for digging deeper and iterating @jiqing-feng ! 💛 

Regarding tests: it's a bit tricky to test two devices on our CI AFAIK 🤔 @amyeroberts do you have suggestions on how to test it? [TL;DR @jiqing-feng found that assisted generation fails if the two models are on different devices, because the special tokens are copied from the main model to the assistant model]

**Comment by jiqing-feng** [comment]
> This makes sense, thank you for digging deeper and iterating @jiqing-feng ! 💛
> 
> Regarding tests: it's a bit tricky to test two devices on our CI AFAIK 🤔 @amyeroberts do you have suggestions on how to test it? [TL;DR @jiqing-feng found that assisted generation fails if the two models are on different devices, because the special tokens are copied from the main model to the assistant model]

I think we can just run the test on a device with GPU; there is almost no limitation for CPU because we can run a very tiny model on CPU just for functionality.

**Comment by amyeroberts** [comment]
> Regarding tests: it's a bit tricky to test two devices on our CI AFAIK 🤔 @amyeroberts do you have suggestions on how to test it? [TL;DR @jiqing-feng found that assisted generation fails if the two models are on different devices, because the special tokens are copied from the main model to the assistant model]

@gante There's certain tests in our suite which require multiple devices e.g. [test_model_parallelization](https://github.com/huggingface/transformers/blob/1f9387d33d40562add4be03176e22ddf9ed0d4ab/tests/test_modeling_common.py#L2911), which we can denote with the `require_torch_multi_accelerator` and `require_torch_multi_gpu` decorators. 

In this case, I'd suggest having two tests, one for the single accelerator case, and another which only runs in the multi device case.

**Comment by gante** [comment]
derp, ofc a GPU is enough (which is has a CPU paired up), what a brain fart on my end :D

@jiqing-feng could you two tests like the script in [this comment of yours](https://github.com/huggingface/transformers/pull/31401#issuecomment-2171993325) to [this file](https://github.com/huggingface/transformers/blob/main/tests/generation/test_utils.py)? More precisely:
1. Inside `GenerationIntegrationTests`;
2. Using the `@slow` decorator;
3. One of the tests with the `@require_torch_multi_gpu` decorator with each model in a different gpu, another with `@require_torch_gpu` with the assistant on cpu
4. Let's use one of our tiny test models like `hf-internal-testing/tiny-random-MistralForCausalLM` (as both main model and assistant)

**Comment by jiqing-feng** [comment]
Hi @gante . I have added the tests, could you please take a review? Thx!

**Comment by jiqing-feng** [comment]
Hi @amyeroberts. Could you please take a review? The failed CIs are not related to my changes :)

**Comment by amyeroberts** [comment]
@jiqing-feng Regarding the failing tests, could you rebase on main to include upstream changes? This should resolve the failures on CI 

Could you also run and share the output of executing the following in a multi-gpu environment:  

```
RUN_SLOW=1 pytest -k "test_assisted_decoding_in_different_gpu or test_assisted_decoding_in_different_gpu"
```

**Comment by gante** [comment]
@jiqing-feng rebasing the PR should get CI green 🤗

**Comment by jiqing-feng** [comment]
Hi @amyeroberts . I run the 2 tests individually and got passed, see
image


I also run your command and got the following output
image
These failed tests are due to some import error:
image

**Comment by jiqing-feng** [comment]
Hi @amyeroberts . Do you need more actions before merging? Please let me know, thx!

**Comment by jiqing-feng** [comment]
Hi @amyeroberts @gante . I think this PR should be ready to merge :)

**Comment by amyeroberts** [comment]
@jiqing-feng OK, sorry, I think I messed up with the pytest command. Could you try this instead: 

```
RUN_SLOW=1 pytest tests/generation/test_utils.py::GenerationIntegrationTests::test_assisted_decoding_in_different_gpu
RUN_SLOW=1 pytest tests/generation/test_utils.py::GenerationIntegrationTests::test_assisted_decoding_in_gpu_cpu 
```

**Comment by jiqing-feng** [comment]
Hi @amyeroberts . The failed CIs are not relate to my changes, would you please review my changes?

**Comment by jiqing-feng** [comment]
Hi @amyeroberts @gante , would you please help to merge this PR? Thx!

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

**Comment by amyeroberts** [comment]
Hi @jiqing-feng, we had to wait for somethings to be resolved upstream and to wait for a new CI run (which I triggered last night)

**Pull request merged**
